pkg/controller/certificatesigningrequest/ipsec_csr_signing_controller.go

```diff
@@ -0,0 +1,486 @@
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		caConfigMap = &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      ipsecRootCAName,
+				Namespace: env.GetAntreaNamespace(),
+			},
+			Data: desiredConfigMapData,
+		}
+		caConfigMap, err = c.client.CoreV1().ConfigMaps(env.GetAntreaNamespace()).Create(context.TODO(), caConfigMap, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+		klog.InfoS("Created ConfigMap for self-signed root CA")
+	}
+	if !reflect.DeepEqual(desiredConfigMapData, caConfigMap.Data) {
+		caConfigMap.Data = desiredConfigMapData
+		_, err = c.client.CoreV1().ConfigMaps(env.GetAntreaNamespace()).Update(context.TODO(), caConfigMap, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *IPsecCSRSigningController) csrWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// watchSecretChanges use watch API directly to watch for Secret changes.
+// Antrea Controller should not have List permission for Secrets.
+func (c *IPsecCSRSigningController) watchSecretChanges(endCh <-chan struct{}) error {
+	// re-queue in case of hitting any errors when creating a new watcher.
+	c.fixturesQueue.Add(workerItemKey)
+
+	watcher, err := c.client.CoreV1().Secrets(env.GetAntreaNamespace()).Watch(context.TODO(), metav1.SingleObject(metav1.ObjectMeta{
+		Namespace: env.GetAntreaNamespace(),
+		Name:      ipsecRootCAName,
+	}))
+	if err != nil {
+		return fmt.Errorf("failed to create Secret watcher: %v", err)
+	}
+	// re-queue in case of missing events before watcher starts.
+	c.fixturesQueue.Add(workerItemKey)
+	ch := watcher.ResultChan()
+	defer watcher.Stop()
+	for {
+		select {
+		case _, ok := <-ch:
+			if !ok {
+				return fmt.Errorf("event channel closed")
+			}
+			// we do not care the actual Event.
+			c.fixturesQueue.Add(workerItemKey)
+		case <-endCh:
+			return nil
+		}
+	}
+}
+
+func (c *IPsecCSRSigningController) fixturesWorker() {
+	for c.processNextFixtureWorkItem() {
+	}
+}
+
+func (c *IPsecCSRSigningController) enqueueCertificateSigningRequest(obj interface{}) {
+	csr, ok := obj.(*certificatesv1.CertificateSigningRequest)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(nil, "Received unexpected object", "obj", obj)
+			return
+		}
+		csr, ok = deletedState.Obj.(*certificatesv1.CertificateSigningRequest)
+		if !ok {
+			klog.ErrorS(nil, "DeletedFinalStateUnknown contains non-CertificateSigningRequest object", "object", deletedState.Obj)
+			return
+		}
+	}
+	c.queue.Add(csr.Name)
+}
+
+func (c *IPsecCSRSigningController) syncCSR(key string) error {
+	startTime := time.Now()
+	defer func() {
+		d := time.Since(startTime)
+		klog.V(2).InfoS("Finished syncing CertificateSigningRequest", "name", key, "duration", d)
+	}()
+	csr, err := c.csrLister.Get(key)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	if csr.Spec.SignerName != antreaapis.AntreaIPsecCSRSignerName {
```

Comment:
approvingController and signingController should specify a filter based on `spec.signerName` field when starting the watch to reduce overhead on both server and clients.

Reply:
added filter.